### ISSUE TITLE
Feature Creap: Manual Save on Idle

### ISSUE THREAD

**Issue by comods** [Gruntfuggly/formatOnIdle#3]
TSLint already formats on manual saves. I am using another extention that triggers a manual save on keypress (Called: Save Typing). If it isn't too much trouble, your extention having a config option to also trigger a manual save on idle would be a better solution then each keypress (The other extention breaks undo). This would trigger most custom formatters without any extra configuration.

**Comment by Gruntfuggly** [comment]
Sounds like a different extension, but sharing most of the code. Probably just 'do stuff on idle' :grin:

It might be simpler to rename this one.

Regarding the undo, I'm pretty sure formatOnIdle breaks it too - i.e. if you undo a few keypresses and the format then gets triggered, you'll lost the redo history.

Unfortunately there doesn't seem to be way of detecting an undo event - there is a document version number that gets updated with each change, but that still increments when you undo. It could potentially store hashes of the whole document and only trigger the format if they don't match. Not sure what the performance would be like though...

**Comment by comods** [comment]
What I think is happening in the other extention:

(1) Manual-Press-Tab -> Auto-Save -> Auto-Format (Removes Tab) -> Auto-Save -> Auto-Format (No changes) ->
(2) Manual-Undo-Action (Undoes Tab Removal) -> Auto-Save -> Auto-Format (Redoes Tab Removal) -> Auto-Save -> Auto-Format (No changes) ->
(3) Undo history will get stuck repeating "(2)" unable to undo further back.

Which means that just pressing undo faster than your idle timeout would fix the issue. Since the other extension is based on keypress, it causes the above loop.

Could you see if you can capture the key combo that triggers undo? If you can at least see the keys being pressed, then you could have a config option list of key-combos that disable your idle timeout. (Defaulting to the default undo keycombo) This could also be useful to not trigger your extetion after a manual Ctrl-S.

Thanks again for your extention, VSCode shouldn't have disabled this functionally (Auto Save Delay doesn't format) but instead put it behind a config option.

**Comment by comods** [comment]
"TriggerOnIdle" "DoOnIdle" "CallOnIdle" "CustomOnIdle" "AnythingOnIdle"

**Comment by Gruntfuggly** [comment]
I don't think you can find out what keys were pressed - you just get an event which tells you the document was changed and a field saying if it was by Keyboard, Mouse or Command.

I'll experiment with hashing the document contents...

**Comment by comods** [comment]
A simple bandaid is, if it was changed by Command, then double or tripple your timeout. That way someone has time to press the undo button multiple times.

**Comment by Gruntfuggly** [comment]
Yeh, or maybe ignore 'command' changes altogther...

**Comment by comods** [comment]
What if you create another extention that triggers on select commands, then forwards that info to this extention:

"activationEvents": [
    "onCommand:extension.sayHello"
]

but on: 
undo Command

**Comment by comods** [comment]
There is a Soft Undo (just moves cursor), which you might be able to save current cursor location, then use Soft Undo, and use this last undo location like a hash, then move the cursor back to normal before formating.

**Comment by Gruntfuggly** [comment]
I've just uploaded a new version which inhibits the format on undo/redo. Can you let me know if it works OK for you ? If it does, I'll make the mods to add more commands...

**Comment by Gruntfuggly** [comment]
I've just created a new extension based on this called 'On Idle'. It's pretty crude, it just runs the configured commands on idle. It has some obvious shortcomings:
- no parameters
- no capability to wait for a command to finish

Let me know what you think...
